src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs

```diff
@@ -228,31 +222,49 @@ await scope
                 await recipeExecutor.ExecuteAsync(executionId, context.Recipe, context.Properties, _applicationLifetime.ApplicationStopping);
             }
 
-            // Reloading the shell context as the recipe has probably updated its features
+            if (context.Errors.Count > 0)
+            {
+                return executionId;
+            }
+
+            // Reloading the shell context as the recipe has probably updated its features.
             await (await _shellHost.GetScopeAsync(shellSettings)).UsingAsync(async scope =>
             {
+                var logger = scope.ServiceProvider.GetRequiredService<ILogger<SetupService>>();
+                var handlers = scope.ServiceProvider.GetServices<ITenantSetupEventHandler>();
+
+                await handlers.InvokeAsync((handler, ctx) => handler.SetupAsync(ctx), context, _logger);
+
+#pragma warning disable CS0618 // Type or member is obsolete
+                // Invoke modules to react to the setup event.
+                var setupEventHandlers = scope.ServiceProvider.GetServices<ISetupEventHandler>();
+#pragma warning restore CS0618 // Type or member is obsolete
+
                 void reportError(string key, string message)
                 {
                     context.Errors[key] = message;
                 }
 
-                // Invoke modules to react to the setup event
-                var setupEventHandlers = scope.ServiceProvider.GetServices<ISetupEventHandler>();
-                var logger = scope.ServiceProvider.GetRequiredService<ILogger<SetupService>>();
+                await setupEventHandlers.InvokeAsync((handler, ctx) => handler.Setup(ctx.Properties, reportError), context, logger);
 
-                await setupEventHandlers.InvokeAsync((handler, context) => handler.Setup(
-                    context.Properties,
-                    reportError
-                ), context, logger);
+                if (context.Errors.Count > 0)
+                {
+                    await handlers.InvokeAsync((handler) => handler.FailedAsync(context), _logger);
+                }
             });
 
-            if (context.Errors.Any())
+            if (context.Errors.Count == 0)
             {
-                return executionId;
```

Comment:
I like the previous pattern, first check the opposite and return if true, as aspnetcore often do (but not always), it's more readable and prevent indentation afterwards.

But yes here not so important, just to keep the code as it was before as much as possible.